flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectResultFetcher.java

```diff
@@ -0,0 +1,343 @@
+		return (CollectCoordinationResponse) gateway.sendCoordinationRequest(operatorId, request).get();
+	}
+
+	@Nullable
+	private Tuple2<Long, CollectCoordinationResponse> getAccumulatorResults() {
+		checkJobClientConfigured();
+
+		JobExecutionResult executionResult;
+		try {
+			// this timeout is sort of hack, see comments in isJobTerminated for explanation
+			executionResult = jobClient.getJobExecutionResult(getClass().getClassLoader()).get(
+				DEFAULT_ACCUMULATOR_GET_MILLIS, TimeUnit.MILLISECONDS);
+		} catch (InterruptedException | ExecutionException | TimeoutException e) {
+			throw new RuntimeException("Failed to fetch job execution result", e);
+		}
+
+		ArrayList<byte[]> accResults = executionResult.getAccumulatorResult(accumulatorName);
+		if (accResults == null) {
+			// job terminates abnormally
+			return null;
+		}
+
+		try {
+			List<byte[]> serializedResults =
+				SerializedListAccumulator.deserializeList(accResults, BytePrimitiveArraySerializer.INSTANCE);
+			byte[] serializedResult = serializedResults.get(0);
+			return CollectSinkFunction.deserializeAccumulatorResult(serializedResult);
+		} catch (ClassNotFoundException | IOException e) {
+			// this is impossible
+			throw new RuntimeException("Failed to deserialize accumulator results", e);
+		}
+	}
+
+	private boolean isJobTerminated() {
+		checkJobClientConfigured();
+
+		try {
+			JobStatus status = jobClient.getJobStatus().get();
+			return status.isGloballyTerminalState();
+		} catch (Exception e) {
+			// TODO
+			//  This is sort of hack.
+			//  Currently different execution environment will have different behaviors
+			//  when fetching a finished job status.
+			//  For example, standalone session cluster will return a normal FINISHED,
+			//  while mini cluster will throw IllegalStateException,
+			//  and yarn per job will throw ApplicationNotFoundException.
+			//  We have to assume that job has finished in this case.
+			//  Change this when these behaviors are unified.
+			LOG.warn("Failed to get job status so we assume that the job has terminated. Some data might be lost.", e);
+			return true;
+		}
+	}
+
+	private void cancelJob() {
+		checkJobClientConfigured();
+
+		if (!isJobTerminated()) {
+			jobClient.cancel();
+		}
+	}
+
+	private void sleepBeforeRetry() {
+		if (retryMillis <= 0) {
+			return;
+		}
+
+		try {
+			// TODO a more proper retry strategy?
+			Thread.sleep(retryMillis);
+		} catch (InterruptedException e) {
+			LOG.warn("Interrupted when sleeping before a retry", e);
+		}
+	}
+
+	private void checkJobClientConfigured() {
+		Preconditions.checkNotNull(jobClient, "Job client must be configured before first use.");
+	}
+
+	private class ResultBuffer {
+
+		private static final String INIT_VERSION = "";
+
+		private final LinkedList<T> buffer;
+		private final TypeSerializer<T> serializer;
+
+		private String version;
+		private long offset;
+		private long lastCheckpointedOffset;
+		private long userHead;
```

Comment:
We need some document about these variables and ResultBuffer, I'm quite confused now even if I know the algorithm...